File: python/net.py
"""The AlphaZero policy/value network on the self-made NumPy autograd.

Python twin of net.cpp: a small conv ResNet with two heads (policy logits over
A = n*n+1 actions incl. pass; value scalar in (-1,1) via tanh). Same Conv->BN->ReLU
block reused from the from-scratch YOLO autograd.
"""

import math
import pickle
import numpy as np

import autograd as ag
from autograd import Tensor
from game import Othello


class ConvBN:
    """Conv (no bias) -> BatchNorm -> (ReLU applied by `block`). He-style init."""

    def __init__(self, cin, cout, k, stride, pad):
        self.w = Tensor.randn((cout, cin, k, k), math.sqrt(2.0 / (cin * k * k)), True)
        self.bias = Tensor.zeros((cout,), False)     # fixed zero; BN.beta shifts
        self.gamma = Tensor(np.ones(cout), requires_grad=True)
        self.beta = Tensor.zeros((cout,), requires_grad=True)
        self.run_mean = np.zeros(cout)               # BN buffers (no grad)
        self.run_var = np.ones(cout)
        self.stride, self.pad = stride, pad

    def params(self):
        return [self.w, self.gamma, self.beta]


def _cbn(net, c, x, act):
    h = ag.conv2d(x, c.w, c.bias, c.stride, c.pad)
    h = ag.batchnorm2d(h, c.gamma, c.beta, c.run_mean, c.run_var, net.training)
    return ag.relu(h) if act else h


def _fc(x, w, b):
    return ag.add_bias_2d(ag.matmul(x, w), b)        # (N,in)@(in,out)+ (out)


def _fc_weight(fin, fout):
    return Tensor.randn((fin, fout), math.sqrt(2.0 / fin), True)


class ResBlock:
    def __init__(self, ch):
        self.a = ConvBN(ch, ch, 3, 1, 1)
        self.b = ConvBN(ch, ch, 3, 1, 1)


class Net:
    def __init__(self, boardn, channels=32, nblocks=3, vhid=32):
        self.n, self.ch, self.nblocks, self.vhid = boardn, channels, nblocks, vhid
        self.hw = boardn * boardn
        self.A = self.hw + 1
        self.training = True

        self.stem = ConvBN(3, channels, 3, 1, 1)
        self.blocks = [ResBlock(channels) for _ in range(nblocks)]
        self.phead = ConvBN(channels, 2, 1, 1, 0)
        self.pw = _fc_weight(2 * self.hw, self.A)
        self.pb = Tensor.zeros((self.A,), True)
        self.vhead = ConvBN(channels, 1, 1, 1, 0)
        self.vw1 = _fc_weight(self.hw, vhid)
        self.vb1 = Tensor.zeros((vhid,), True)
        self.vw2 = _fc_weight(vhid, 1)
        self.vb2 = Tensor.zeros((1,), True)

    def forward(self, x):
        N = x.shape[0]
        h = _cbn(self, self.stem, x, True)
        for blk in self.blocks:
            y = _cbn(self, blk.a, h, True)
            y = _cbn(self, blk.b, y, False)
            h = ag.relu(ag.add(h, y))                 # residual
        # policy head
        p = _cbn(self, self.phead, h, True)           # (N,2,n,n)
        p = ag.reshape(p, (N, 2 * self.hw))
        logits = _fc(p, self.pw, self.pb)             # (N,A)
        # value head
        v = _cbn(self, self.vhead, h, True)           # (N,1,n,n)
        v = ag.reshape(v, (N, self.hw))
        v = ag.relu(_fc(v, self.vw1, self.vb1))       # (N,vhid)
        v = ag.tanh_(_fc(v, self.vw2, self.vb2))      # (N,1)
        return logits, v

    def params(self):
        p = []
        p += self.stem.params()
        for blk in self.blocks:
            p += blk.a.params() + blk.b.params()
        p += self.phead.params() + [self.pw, self.pb]
        p += self.vhead.params() + [self.vw1, self.vb1, self.vw2, self.vb2]
        return p

    def buffers(self):
        b = []
        b += [self.stem.run_mean, self.stem.run_var]
        for blk in self.blocks:
            b += [blk.a.run_mean, blk.a.run_var, blk.b.run_mean, blk.b.run_var]
        b += [self.phead.run_mean, self.phead.run_var, self.vhead.run_mean, self.vhead.run_var]
        return b

    def loss(self, x, target_pi, target_z):
        N = x.shape[0]
        logits, v = self.forward(x)
        logp = ag.log_softmax_rows(logits)
        # policy cross-entropy: -(1/N) sum_n sum_a pi * log_softmax
        pol = ag.mul_scalar(ag.sum(ag.mul(target_pi, logp)), -1.0 / N)
        diff = ag.sub(v, target_z)                    # value MSE
        val = ag.mean(ag.mul(diff, diff))
        return ag.add(pol, val)

    def eval_state(self, s):
        """Batch-1 inference for MCTS: returns (logits over A, value scalar)."""
        prev = self.training
        self.training = False
        x = Tensor(s.encode()[None])                  # (1,3,n,n)
        logits, v = self.forward(x)
        self.training = prev
        return logits.data[0].copy(), float(v.data[0, 0])

    def copy_from(self, other):
        for p, q in zip(self.params(), other.params()):
            p.data[...] = q.data
        for b, c in zip(self.buffers(), other.buffers()):
            b[...] = c

    def config(self):
        return dict(n=self.n, ch=self.ch, nblocks=self.nblocks, vhid=self.vhid)

    def save(self, path):
        with open(path, "wb") as f:
            pickle.dump({"config": self.config(),
                         "params": [p.data for p in self.params()],
                         "buffers": [b for b in self.buffers()]}, f)

    def load(self, path):
        with open(path, "rb") as f:
            d = pickle.load(f)
        for p, arr in zip(self.params(), d["params"]):
            p.data[...] = arr
        for b, arr in zip(self.buffers(), d["buffers"]):
            b[...] = arr

    @staticmethod
    def from_file(path):
        with open(path, "rb") as f:
            cfg = pickle.load(f)["config"]
        net = Net(cfg["n"], cfg["ch"], cfg["nblocks"], cfg["vhid"])
        net.load(path)
        net.training = False
        return net

File: net.cpp
#include "net.h"
#include <cmath>
#include <cstdio>

using namespace ag;

namespace oth {

static Tensor ones(std::vector<int> shape, bool rg) {
    auto t = Tensor::zeros(std::move(shape), rg);
    for (auto& v : t.data()) v = 1.f;
    return t;
}

static ConvBN make_convbn(int cin, int cout, int k, int stride, int pad) {
    ConvBN c;
    c.w = Tensor::randn({cout, cin, k, k}, std::sqrt(2.0f / (cin * k * k)), true);
    c.bias = Tensor::zeros({cout}, false);
    c.gamma = ones({cout}, true);
    c.beta = Tensor::zeros({cout}, true);
    c.run_mean = Tensor::zeros({cout}, false);
    c.run_var = ones({cout}, false);
    c.stride = stride; c.pad = pad;
    return c;
}

// Conv -> BatchNorm -> [ReLU]
static Tensor cbn(Net& net, ConvBN& c, const Tensor& x, bool act) {
    auto h = conv2d(x, c.w, c.bias, c.stride, c.pad);
    h = batchnorm2d(h, c.gamma, c.beta, c.run_mean, c.run_var, net.training);
    return act ? relu(h) : h;
}

// FC: (N,in) @ W(in,out) + b(out)
static Tensor fc(const Tensor& x, const Tensor& w, const Tensor& b) {
    return add_bias_2d(matmul(x, w), b);
}

static Tensor fc_weight(int fin, int fout) {
    return Tensor::randn({fin, fout}, std::sqrt(2.0f / fin), true);
}

Net::Net(int boardn, int channels, int nb, int valhid)
    : n(boardn), ch(channels), nblocks(nb), vhid(valhid) {
    hw = n * n;
    A = hw + 1;
    stem = make_convbn(3, ch, 3, 1, 1);
    for (int i = 0; i < nblocks; ++i)
        blocks.push_back({make_convbn(ch, ch, 3, 1, 1), make_convbn(ch, ch, 3, 1, 1)});
    phead = make_convbn(ch, 2, 1, 1, 0);
    pw = fc_weight(2 * hw, A);
    pb = Tensor::zeros({A}, true);
    vhead = make_convbn(ch, 1, 1, 1, 0);
    vw1 = fc_weight(hw, vhid);
    vb1 = Tensor::zeros({vhid}, true);
    vw2 = fc_weight(vhid, 1);
    vb2 = Tensor::zeros({1}, true);
}

std::pair<Tensor, Tensor> Net::forward(const Tensor& x) {
    int N = x.shape()[0];
    auto h = cbn(*this, stem, x, true);
    for (auto& blk : blocks) {
        auto y = cbn(*this, blk.a, h, true);
        y = cbn(*this, blk.b, y, false);
        h = relu(add(h, y));                       // residual
    }
    // policy head
    auto p = cbn(*this, phead, h, true);           // (N,2,n,n)
    p = reshape(p, {N, 2 * hw});
    auto logits = fc(p, pw, pb);                    // (N,A)
    // value head
    auto v = cbn(*this, vhead, h, true);           // (N,1,n,n)
    v = reshape(v, {N, hw});
    v = relu(fc(v, vw1, vb1));                       // (N,vhid)
    v = tanh_(fc(v, vw2, vb2));                      // (N,1)
    return {logits, v};
}

std::vector<Tensor> Net::params() {
    std::vector<Tensor> p;
    auto add_cbn = [&](ConvBN& c) { p.push_back(c.w); p.push_back(c.gamma); p.push_back(c.beta); };
    add_cbn(stem);
    for (auto& blk : blocks) { add_cbn(blk.a); add_cbn(blk.b); }
    add_cbn(phead); p.push_back(pw); p.push_back(pb);
    add_cbn(vhead); p.push_back(vw1); p.push_back(vb1); p.push_back(vw2); p.push_back(vb2);
    return p;
}

Tensor Net::loss(const Tensor& x, const Tensor& target_pi, const Tensor& target_z) {
    int N = x.shape()[0];
    auto pr = forward(x);
    auto logp = log_softmax_rows(pr.first);
    // policy cross-entropy: -(1/N) sum_n sum_a pi * log_softmax
    auto pol = mul_scalar(sum(mul(target_pi, logp)), -1.0f / N);
    // value MSE
    auto diff = sub(pr.second, target_z);
    auto val = mean(mul(diff, diff));
    return add(pol, val);
}

std::vector<Tensor> Net::buffers() {
    std::vector<Tensor> b;
    auto add_cbn = [&](ConvBN& c) { b.push_back(c.run_mean); b.push_back(c.run_var); };
    add_cbn(stem);
    for (auto& blk : blocks) { add_cbn(blk.a); add_cbn(blk.b); }
    add_cbn(phead); add_cbn(vhead);
    return b;
}

void Net::copy_from(Net& other) {
    auto ps = params(), qs = other.params();
    for (size_t i = 0; i < ps.size(); ++i) ps[i].data() = qs[i].data();
    auto bs = buffers(), cs = other.buffers();
    for (size_t i = 0; i < bs.size(); ++i) bs[i].data() = cs[i].data();
}

bool Net::save(const std::string& path) {
    FILE* f = std::fopen(path.c_str(), "wb");
    if (!f) return false;
    auto write = [&](std::vector<Tensor> ts) {
        for (auto& t : ts) {
            int nsz = t.numel();
            std::fwrite(&nsz, sizeof(int), 1, f);
            std::fwrite(t.data().data(), sizeof(float), nsz, f);
        }
    };
    write(params()); write(buffers());
    std::fclose(f);
    return true;
}

bool Net::load(const std::string& path) {
    FILE* f = std::fopen(path.c_str(), "rb");
    if (!f) return false;
    auto read = [&](std::vector<Tensor> ts) {
        for (auto& t : ts) {
            int nsz = 0;
            if (std::fread(&nsz, sizeof(int), 1, f) != 1 || nsz != t.numel()) return false;
            std::fread(t.data().data(), sizeof(float), nsz, f);
        }
        return true;
    };
    bool ok = read(params()) && read(buffers());
    std::fclose(f);
    return ok;
}

void Net::eval_state(const Othello& s, std::vector<float>& logits, float& value) {
    // NOTE: does not touch `training` (that would be a data race under parallel
    // self-play). Callers must put the net in eval mode first (training=false),
    // which self-play / arena / eval_vs_random all do.
    std::vector<float> planes;
    s.encode(planes);
    auto x = Tensor::from(std::move(planes), {1, 3, n, n}, false);
    auto pr = forward(x);
    logits.assign(pr.first.data().begin(), pr.first.data().end());   // A values
    value = pr.second.data()[0];
}

}  // namespace oth
